File: src/Other/RelayCommand.cs
using System;
using System.Runtime.CompilerServices;
using System.Windows.Input;

namespace OpenSilver.Samples.Showcase;

/// <summary>
/// An interface expanding <see cref="ICommand"/> with the ability to raise
/// the <see cref="ICommand.CanExecuteChanged"/> event externally.
/// </summary>
public interface IRelayCommand : ICommand
{
    /// <summary>
    /// Notifies that the <see cref="ICommand.CanExecute"/> property has changed.
    /// </summary>
    void NotifyCanExecuteChanged();
}

/// <summary>
/// A command whose sole purpose is to relay its functionality to other
/// objects by invoking delegates. The default return value for the <see cref="CanExecute"/>
/// method is <see langword="true"/>. This type does not allow you to accept command parameters
/// in the <see cref="Execute"/> and <see cref="CanExecute"/> callback methods.
/// </summary>
public sealed class RelayCommand : IRelayCommand
{
    /// <summary>
    /// The <see cref="Action"/> to invoke when <see cref="Execute"/> is used.
    /// </summary>
    private readonly Action _execute;

    /// <summary>
    /// The optional action to invoke when <see cref="CanExecute"/> is used.
    /// </summary>
    private readonly Func<bool> _canExecute;

    /// <inheritdoc/>
    public event EventHandler CanExecuteChanged;

    /// <summary>
    /// Initializes a new instance of the <see cref="RelayCommand"/> class that can always execute.
    /// </summary>
    /// <param name="execute">The execution logic.</param>
    /// <exception cref="ArgumentNullException">Thrown if <paramref name="execute"/> is <see langword="null"/>.</exception>
    public RelayCommand(Action execute)
    {
        _execute = execute ?? throw new ArgumentNullException(nameof(execute));
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="RelayCommand"/> class.
    /// </summary>
    /// <param name="execute">The execution logic.</param>
    /// <param name="canExecute">The execution status logic.</param>
    /// <exception cref="ArgumentNullException">Thrown if <paramref name="execute"/> or <paramref name="canExecute"/> are <see langword="null"/>.</exception>
    public RelayCommand(Action execute, Func<bool> canExecute)
    {
        _execute = execute ?? throw new ArgumentNullException(nameof(execute)); ;
        _canExecute = canExecute ?? throw new ArgumentNullException(nameof(canExecute)); ;
    }

    /// <inheritdoc/>
    public void NotifyCanExecuteChanged()
    {
        CanExecuteChanged?.Invoke(this, EventArgs.Empty);
    }

    /// <inheritdoc/>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool CanExecute(object parameter)
    {
        return _canExecute?.Invoke() != false;
    }

    /// <inheritdoc/>
    public void Execute(object parameter)
    {
        _execute();
    }
}

/// <summary>
/// A generic interface representing a more specific version of <see cref="IRelayCommand"/>.
/// </summary>
/// <typeparam name="T">The type used as argument for the interface methods.</typeparam>
public interface IRelayCommand<in T> : IRelayCommand
{
    /// <summary>
    /// Provides a strongly-typed variant of <see cref="ICommand.CanExecute(object)"/>.
    /// </summary>
    /// <param name="parameter">The input parameter.</param>
    /// <returns>Whether or not the current command can be executed.</returns>
    /// <remarks>Use this overload to avoid boxing, if <typeparamref name="T"/> is a value type.</remarks>
    bool CanExecute(T parameter);

    /// <summary>
    /// Provides a strongly-typed variant of <see cref="ICommand.Execute(object)"/>.
    /// </summary>
    /// <param name="parameter">The input parameter.</param>
    /// <remarks>Use this overload to avoid boxing, if <typeparamref name="T"/> is a value type.</remarks>
    void Execute(T parameter);
}

/// <summary>
/// A generic command whose sole purpose is to relay its functionality to other
/// objects by invoking delegates. The default return value for the CanExecute
/// method is <see langword="true"/>. This class allows you to accept command parameters
/// in the <see cref="Execute(T)"/> and <see cref="CanExecute(T)"/> callback methods.
/// </summary>
/// <typeparam name="T">The type of parameter being passed as input to the callbacks.</typeparam>
public sealed class RelayCommand<T> : IRelayCommand<T>
{
    /// <summary>
    /// The <see cref="Action"/> to invoke when <see cref="Execute(T)"/> is used.
    /// </summary>
    private readonly Action<T> _execute;

    /// <summary>
    /// The optional action to invoke when <see cref="CanExecute(T)"/> is used.
    /// </summary>
    private readonly Predicate<T> _canExecute;

    /// <inheritdoc/>
    public event EventHandler CanExecuteChanged;

    /// <summary>
    /// Initializes a new instance of the <see cref="RelayCommand{T}"/> class that can always execute.
    /// </summary>
    /// <param name="execute">The execution logic.</param>
    /// <remarks>
    /// Due to the fact that the <see cref="ICommand"/> interface exposes methods that accept a
    /// nullable <see cref="object"/> parameter, it is recommended that if <typeparamref name="T"/> is a reference type,
    /// you should always declare it as nullable, and to always perform checks within <paramref name="execute"/>.
    /// </remarks>
    /// <exception cref="ArgumentNullException">Thrown if <paramref name="execute"/> is <see langword="null"/>.</exception>
    public RelayCommand(Action<T> execute)
    {
        _execute = execute ?? throw new ArgumentNullException(nameof(execute));
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="RelayCommand{T}"/> class.
    /// </summary>
    /// <param name="execute">The execution logic.</param>
    /// <param name="canExecute">The execution status logic.</param>
    /// <remarks>See notes in <see cref="RelayCommand{T}(Action{T})"/>.</remarks>
    /// <exception cref="ArgumentNullException">Thrown if <paramref name="execute"/> or <paramref name="canExecute"/> are <see langword="null"/>.</exception>
    public RelayCommand(Action<T> execute, Predicate<T> canExecute)
    {
        _execute = execute ?? throw new ArgumentNullException(nameof(execute));
        _canExecute = canExecute ?? throw new ArgumentNullException(nameof(canExecute)); ;
    }

    /// <inheritdoc/>
    public void NotifyCanExecuteChanged()
    {
        CanExecuteChanged?.Invoke(this, EventArgs.Empty);
    }

    /// <inheritdoc/>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public bool CanExecute(T parameter)
    {
        return _canExecute?.Invoke(parameter) != false;
    }

    /// <inheritdoc/>
    public bool CanExecute(object parameter)
    {
        // Special case a null value for a value type argument type.
        // This ensures that no exceptions are thrown during initialization.
        if (parameter is null && default(T) is not null)
        {
            return false;
        }

        if (!TryGetCommandArgument(parameter, out T result))
        {
            ThrowArgumentExceptionForInvalidCommandArgument(parameter);
        }

        return CanExecute(result);
    }

    /// <inheritdoc/>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void Execute(T parameter)
    {
        _execute(parameter);
    }

    /// <inheritdoc/>
    public void Execute(object parameter)
    {
        if (!TryGetCommandArgument(parameter, out T result))
        {
            ThrowArgumentExceptionForInvalidCommandArgument(parameter);
        }

        Execute(result);
    }

    /// <summary>
    /// Tries to get a command argument of compatible type <typeparamref name="T"/> from an input <see cref="object"/>.
    /// </summary>
    /// <param name="parameter">The input parameter.</param>
    /// <param name="result">The resulting <typeparamref name="T"/> value, if any.</param>
    /// <returns>Whether or not a compatible command argument could be retrieved.</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static bool TryGetCommandArgument(object parameter, out T result)
    {
        // If the argument is null and the default value of T is also null, then the
        // argument is valid. T might be a reference type or a nullable value type.
        if (parameter is null && default(T) is null)
        {
            result = default;

            return true;
        }

        // Check if the argument is a T value, so either an instance of a type or a derived
        // type of T is a reference type, an interface implementation if T is an interface,
        // or a boxed value type in case T was a value type.
        if (parameter is T argument)
        {
            result = argument;

            return true;
        }

        result = default;

        return false;
    }

    /// <summary>
    /// Throws an <see cref="ArgumentException"/> if an invalid command argument is used.
    /// </summary>
    /// <param name="parameter">The input parameter.</param>
    /// <exception cref="ArgumentException">Thrown with an error message to give info on the invalid parameter.</exception>
    internal static void ThrowArgumentExceptionForInvalidCommandArgument(object parameter)
    {
        [MethodImpl(MethodImplOptions.NoInlining)]
        static Exception GetException(object parameter)
        {
            if (parameter is null)
            {
                return new ArgumentException($"Parameter \"{nameof(parameter)}\" (object) must not be null, as the command type requires an argument of type {typeof(T)}.", nameof(parameter));
            }

            return new ArgumentException($"Parameter \"{nameof(parameter)}\" (object) cannot be of type {parameter.GetType()}, as the command type requires an argument of type {typeof(T)}.", nameof(parameter));
        }

        throw GetException(parameter);
    }
}

File: src/Other/RelayCommand.vb
Imports System.Runtime.CompilerServices
Imports System.Windows.Input

Namespace OpenSilver.Samples.Showcase

    Public Interface IRelayCommand
        Inherits ICommand
        Sub NotifyCanExecuteChanged()
    End Interface

    Public NotInheritable Class RelayCommand
        Implements IRelayCommand

        Private ReadOnly _execute As Action
        Private ReadOnly _canExecute As Func(Of Boolean)

        Public Sub New(execute As Action)
            If execute Is Nothing Then Throw New ArgumentNullException(NameOf(execute))
            _execute = execute
        End Sub

        Public Sub New(execute As Action, canExecute As Func(Of Boolean))
            If execute Is Nothing Then Throw New ArgumentNullException(NameOf(execute))
            If canExecute Is Nothing Then Throw New ArgumentNullException(NameOf(canExecute))
            _execute = execute
            _canExecute = canExecute
        End Sub

        Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

        Public Sub NotifyCanExecuteChanged() Implements IRelayCommand.NotifyCanExecuteChanged
            RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
        End Sub

        <MethodImpl(MethodImplOptions.AggressiveInlining)>
        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
            Return If(_canExecute?.Invoke(), True)
        End Function

        Public Sub Execute(parameter As Object) Implements ICommand.Execute
            _execute()
        End Sub
    End Class

    Public Interface IRelayCommand(Of In T)
        Inherits IRelayCommand
        Function CanExecute(parameter As T) As Boolean
        Sub Execute(parameter As T)
    End Interface

    Public NotInheritable Class RelayCommand(Of T)
        Implements IRelayCommand(Of T)

        Private ReadOnly _execute As Action(Of T)
        Private ReadOnly _canExecute As Predicate(Of T)

        Public Sub New(execute As Action(Of T))
            If execute Is Nothing Then Throw New ArgumentNullException(NameOf(execute))
            _execute = execute
        End Sub

        Public Sub New(execute As Action(Of T), canExecute As Predicate(Of T))
            If execute Is Nothing Then Throw New ArgumentNullException(NameOf(execute))
            If canExecute Is Nothing Then Throw New ArgumentNullException(NameOf(canExecute))
            _execute = execute
            _canExecute = canExecute
        End Sub

        Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

        Public Sub NotifyCanExecuteChanged() Implements IRelayCommand.NotifyCanExecuteChanged
            RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
        End Sub

        Public Function CanExecute(parameter As T) As Boolean Implements IRelayCommand(Of T).CanExecute
            Return If(_canExecute?.Invoke(parameter), True)
        End Function

        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
            If parameter Is Nothing AndAlso GetType(T).IsValueType AndAlso Nullable.GetUnderlyingType(GetType(T)) Is Nothing Then
                Return False
            End If

            Dim result As T
            If Not TryGetCommandArgument(parameter, result) Then
                ThrowArgumentExceptionForInvalidCommandArgument(parameter)
            End If

            Return CanExecute(result)
        End Function

        Public Sub Execute(parameter As T) Implements IRelayCommand(Of T).Execute
            _execute(parameter)
        End Sub

        Public Sub Execute(parameter As Object) Implements ICommand.Execute
            Dim result As T
            If Not TryGetCommandArgument(parameter, result) Then
                ThrowArgumentExceptionForInvalidCommandArgument(parameter)
            End If

            Execute(result)
        End Sub

        Private Shared Function TryGetCommandArgument(parameter As Object, ByRef result As T) As Boolean
            If parameter Is Nothing AndAlso GetType(T).IsClass Then
                result = Nothing
                Return True
            End If

            If TypeOf parameter Is T Then
                result = CType(parameter, T)
                Return True
            End If

            result = Nothing
            Return False
        End Function

        Private Shared Sub ThrowArgumentExceptionForInvalidCommandArgument(parameter As Object)
            If parameter Is Nothing Then
                Throw New ArgumentException($"Parameter must not be null. Command requires argument of type {GetType(T).Name}.", NameOf(parameter))
            End If

            Throw New ArgumentException($"Parameter is of type {parameter.GetType().Name}, but expected {GetType(T).Name}.", NameOf(parameter))
        End Sub

    End Class

End Namespace
